internal/dependency/manager.go

```diff
@@ -0,0 +1,268 @@
+	var idx repo2.PackageIndex
+	if err := repo2.FetchPackageIndex(repo, pkgName, &idx); err != nil {
+		return "", err
+	}
+	constraint, err := semver.NewConstraint(versionRange)
+	if err != nil {
+		return "", err
+	}
+	var compatibleVersions []*semver.Version
+	for _, v := range idx.Versions {
+		if version, err := semver.NewVersion(v.Version); err != nil {
+			// TODO
+			continue
+		} else if ok := constraint.Check(version); ok {
+			compatibleVersions = append(compatibleVersions, version)
+		}
+	}
+	if len(compatibleVersions) > 0 {
+		collection := semver.Collection(compatibleVersions)
+		sort.Sort(collection)
+		return collection[len(collection)-1].Original(), nil
+	} else {
+		return "", errors.New("no compatible versions found")
+	}
+}
+
+func NewDependencyManagerForClient(pkgClient *client.PackageV1Alpha1Client) (*DependendcyManager, error) {
+	adapter := &GoClientAdapter{
+		pkgClient: pkgClient,
+	}
+	if err := adapter.Init(); err != nil {
+		return nil, err
+	} else {
+		return &DependendcyManager{
+			clientAdapter: adapter,
+			repoAdapter:   &defaultRepoAdapter{},
+		}, nil
+	}
+}
+
+func NewDependencyManagerForController(client client2.Client) (*DependendcyManager, error) {
+	adapter := &ControllerRuntimeAdapter{
+		client: client,
+	}
+	if err := adapter.Init(); err != nil {
+		return nil, err
+	} else {
+		return &DependendcyManager{
+			clientAdapter: adapter,
+			repoAdapter:   &defaultRepoAdapter{},
+		}, nil
+	}
+}
+
+func (dm *DependendcyManager) CheckRequirements(ctx context.Context, pkg *v1alpha1.Package, pkgInfo *v1alpha1.PackageInfo) (*CheckResult, error) {
+	if pkg == nil || pkgInfo == nil || pkgInfo.Status.Manifest == nil {
+		return nil, errors.New("nil not allowed")
+	}
+	piManifest := pkgInfo.Status.Manifest
+	allPis, err := dm.clientAdapter.GetAllPackageInfos(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var requirements []Requirement
+	var conflicts []Conflict
+	for _, dependency := range piManifest.Dependencies {
+		if reqPkg, err := dm.clientAdapter.GetRequiredPackage(ctx, dependency.Name); err != nil {
+			return nil, err
+		} else if reqPkg == nil {
+			if req, err := dm.createRequirementForMissingDependency(pkg, dependency); err != nil {
+				return nil, err
+			} else {
+				requirements = append(requirements, *req)
+			}
+		} else if dependency.Version != "" {
+			requiredRange, err := semver.NewConstraint(dependency.Version)
+			if err != nil {
+				return nil, err
+			}
+			otherDependencies := dm.getOtherDependencies(allPis, pkgInfo, dependency)
+			if len(otherDependencies) > 0 {
+				for _, otherDep := range otherDependencies {
+					if otherDep.Version == "" {
+						if conflict, err := dm.checkExistingRequiredPackageWithVersionRange(reqPkg, requiredRange, dependency); err != nil {
+							return nil, err
+						} else if conflict != nil {
+							conflicts = append(conflicts, *conflict)
+						}
```

Comment:
As discussed, as it is, this code block can be executed once, outside of the for loop since it does not depend on the loop variable.